dev/membership#18 Enhance parameters for Job.process_membership

mattwire opened the pull request:
Overview
----------------------------------------
This adds some parameters to control the Job.process_membership API/scheduled job per gitlab issues.

https://lab.civicrm.org/dev/membership/issues/18

Before
----------------------------------------
Cannot control behaviour of Job.process_membership.

After
----------------------------------------
Can control Job.process_membership:

- exclude_test_memberships: Exclude test memberships from calculations (default = TRUE)
- only_active_membership_types: Exclude disabled membership types from calculations (default = TRUE)
- exclude_membership_status_ids: Default: Exclude Pending, Cancelled, Expired. Deceased will always be excluded

Technical Details
----------------------------------------

Comments
----------------------------------------
@michaelmcandrew Can you write a unit test to validate the new parameters? There are a number of tests on Job.process_membership in `tests/phpunit/api/v3/JobTest.php` which could be extended/copied.

mattwire committed:
Add parameters to process_membership job

mattwire commented:
@michaelmcandrew @alifrumin

alifrumin commented:
Great @mattwire I am happy to test when you are ready.

michaelmcandrew commented on api/v3/Job.php:
```diff
@@ -470,7 +470,14 @@ function civicrm_api3_job_process_membership($params) {
   }
 
-  $result = CRM_Member_BAO_Membership::updateAllMembershipStatus();
+  // We need to pass this through as a simple array of membership status IDs as values.
```
If the purpose of this code is to accept variations in peoples preferences for passing parameters, I'm not sure I would be so forgiving. Instead I would expect people to pass in parameters in the correct format.

eileenmcnaughton replied:
@mattwire I agree with @michaelmcandrew on this - but the presence of many tests was pushing me to merge it, even though this chuck of code doesn't really make sense to me 

However, I am concerned that we don't declare a type on exclude_membership_status_ids in the metadata and we are not validating it at the api layer & the validation at the next layer is a bit unclear too.

I think that if it were declare as a INT with a pseudoconstant then an array would be accepted but it would be validated

mattwire replied:
@eileenmcnaughton @michaelmcandrew I've added the metadata for the `exclude_membership_status_ids` parameter - it's now validating as an INT but I don't get a pseudoconstant selector in the API3 UI.

michaelmcandrew commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2178,22 +2173,59 @@ public static function getMembershipContributionId($membershipId, $all = FALSE)
+    }
+
+    for ($index = 0; $index < count($params['exclude_membership_status_ids']); $index++) {
+      $queryParams[$index] = [$params['exclude_membership_status_ids'][$index], 'Integer'];
```
Might want to wrap this in an if to cater for the case when you pass an empty array of statuses to exlude

mattwire replied:
@michaelmcandrew This line `$updateCount = $processCount = self::updateDeceasedMembersStatuses();`

michaelmcandrew replied:
ok - makes sense

michaelmcandrew commented on CRM/Member/BAO/Membership.php:
```diff
@@ -2178,22 +2173,59 @@ public static function getMembershipContributionId($membershipId, $all = FALSE)
+        array_search('Deceased', $allStatus),
+      ];
+    }
+    // Deceased is *always* excluded because it is has very specific processing below.
```
I don't see any very specific processing below - am I missing something?

mattwire replied:
@michaelmcandrew This line `$updateCount = $processCount = self::updateDeceasedMembersStatuses();`

michaelmcandrew replied:
OK - makes sense now - thanks for the clarification.

michaelmcandrew commented:
Have added a few comments. Happy to comment some more. Would be easier to review if the status and types stuff was separated.

michaelmcandrew commented:
@elcapo - want to take a look and comment?

elcapo commented:
> @elcapo - want to take a look and comment?

Nothing to add. Looks great to me.

Currently having a problem with loads of "Expired" memberships that `MembershipStatus.calc` properly identifies as "Current" but where I can't use the `process_membership` job to fix them. With this draft, that would be solved, as I'd be able to specify the membership statuses to be excluded.

Thanks @mattwire!

michaelmcandrew commented:
bump @mattwire - do you know what else we need to do to get this merged?

mattwire commented:
@michaelmcandrew thanks - I've just rebased and moved the other bit (disabled memberships) to #17143. Are you / @elcapo able to write some unit tests to cover the new parameters? Then this should be ready to be merged. I've just added a note on which tests could be extended to the description of this PR.

michaelmcandrew commented:
thanks @mattrenner. yes we can do that - @elcapo could you write some tests please? :)

elcapo commented:
> thanks @mattwire. yes we can do that - @elcapo could you write some tests please? :)

Will do! Just let me find some time these days between upgrades.

Commit:
Test the new Job.process.membership parameters

elcapo commented:
@mattwire I've prepared a test suite for the new parameters. See commit [0205a6cb](https://github.com/elcapo/civicrm-core/commit/0205a6cbf238162ac583ac44aab0c43a71536745) in my [membership_disabled](https://github.com/civicrm/civicrm-core/compare/master...elcapo:membership_disabled) branch.

I couldn't see how to make a pull request to your branch, so let me know how you want me to send you the changes.

Also, some tests were failing and I wasn't sure if it was because of the tests, or because of the parameter handling, so I created a second branch, [membership_disabled_fixes](https://github.com/elcapo/civicrm-core/tree/membership_disabled_fixes), that incorporates both the tests and the fixes that made the code pass all the new tests.

elcapo commented:
> @mattwire I've prepared a test suite for the new parameters. See commit [0205a6cb](https://github.com/elcapo/civicrm-core/commit/0205a6cbf238162ac583ac44aab0c43a71536745) in my [membership_disabled](https://github.com/civicrm/civicrm-core/compare/master...elcapo:membership_disabled) branch. I couldn't see how to make a pull request to your branch, so let me know how you want me to send you the changes.

Right after sending my message, I've realized I had a few asserts commented in the last test, so please consider also this additional commit [20d11c61](https://github.com/elcapo/civicrm-core/commit/20d11c61660f7f006d044ea1d0d7a3c53485fe36).

mattwire commented:
Thanks @elcapo I've pulled your commits into this PR and fixed the style warnings. It looks like two of the tests are failing - any ideas?

elcapo commented:
> Thanks @elcapo I've pulled your commits into this PR and fixed the style warnings. It looks like two of the tests are failing - any ideas?

Did you check out my fixes to make the tests pass? I changed a few things like checking if variables were defined in a different way, as the way they were passed as FALSE, sometimes they were ignored and assumed TRUE. Check this https://github.com/elcapo/civicrm-core/commit/7b10697389e0e5b0439eff2e9e89b228308b9b1f and let me know your thoughts about it.

mattwire commented:
> Did you check out my fixes to make the tests pass?
@elcapo I didn't - but now I have and it's still failing on 3 tests - any ideas?

elcapo commented:
> @elcapo I didn't - but now I have and it's still failing on 3 tests - any ideas?

I've just updated the tests so they are a bit easier to use. See https://github.com/elcapo/civicrm-core/commit/68dd0d85d2ecc8d8dfc88dfea5ce550ddbf2b8ad.

I was making several asserts per test and now I'm only making one, so we've now 20 tests that are showing 2 failures.

### Include Test Memberships

When using this call:

```php
civicrm_api3('job', 'process_membership', [
  'exclude_test_memberships' => FALSE,
]);
```

the membership created in `createTestMembershipThatShouldBeCurrent` should be included in the review process (therefore updated to current) but it is still being excluded.

* See [JobProcessMembershipTest.php#L314](https://github.com/elcapo/civicrm-core/blob/68dd0d85d2ecc8d8dfc88dfea5ce550ddbf2b8ad/tests/phpunit/api/v3/JobProcessMembershipTest.php#L314).

### Including Memberships of Inactive Membership Types

When using this call:

```php
$this->callAPISuccess('job', 'process_membership', [
  'only_active_membership_types' => FALSE,
]);
```

the membership created in `createOldMembershipThatShouldBeCurrent` should be included in the review process (therefore updated to current) but it is still being excluded.

* See [JobProcessMembershipTest.php#L373](https://github.com/elcapo/civicrm-core/blob/68dd0d85d2ecc8d8dfc88dfea5ce550ddbf2b8ad/tests/phpunit/api/v3/JobProcessMembershipTest.php#L373).

elcapo commented:
Hi @mattwire,

Is there any problem with the changes I can still help with? I've checked Jenkins and apparently there are some minor style issues:

![Screenshot_2020-05-12 Jenkins](https://user-images.githubusercontent.com/1746692/81695600-c289ce00-945a-11ea-99a0-1eb3cd87f446.png)

Let me know if you want me to fix them.

mattwire commented:
@elcapo Yes please can you fix the style issues

elcapo commented:
@mattwire Please, test again with the file as here: [JobProcessMembershipTest.php](https://github.com/elcapo/civicrm-core/blob/c30c6ea8252afab041b245b22b851af97a08e96e/tests/phpunit/api/v3/JobProcessMembershipTest.php).

There are this two changes:

- [Fix indentation issue](https://github.com/elcapo/civicrm-core/commit/a6652490244f13ea081289e5e0fe82b804b75a3b#diff-fd0cbc8bd8f64679652ca7129e387c84)
- [Fix missing commas in multiline arrays](https://github.com/elcapo/civicrm-core/commit/c30c6ea8252afab041b245b22b851af97a08e96e#diff-fd0cbc8bd8f64679652ca7129e387c84)

I tested the file with *civilint* and it appears to be fine now.

mattwire commented:
@elcapo I'm still getting test failures on this:
```
api_v3_JobProcessMembershipTest.testIncludingTestMembershipsActuallyIncludesThem
api_v3_JobProcessMembershipTest.testIncludingInactiveMembershipTypesConsidersInactive
```
Any ideas? I squashed the commits but should have everything from your branch?

elcapo commented:
> Any ideas? I squashed the commits but should have everything from your branch?

@mattwire Your last commit fixes the issue, right? At least the one of our two tests failing. I've just executed the new tests with your branch and they all worked fine.

![Screenshot from 2020-06-16 12-31-34](https://user-images.githubusercontent.com/1746692/84769453-c1097380-afcd-11ea-9df8-3376401adf5a.png)

Apparently the issue is now with standard membership tests...

![Screenshot_2020-06-16 CiviCRM-Core-PR #34715  Jenkins](https://user-images.githubusercontent.com/1746692/84769493-d67e9d80-afcd-11ea-8999-e97521a4694b.png)

mattwire committed:
Tests should call API not updateAllMembershipStatus() function directly.

mattwire commented:
@elcapo Let's see if the tests pass now

seamuslee001 commented:
Jenkins re test this please

mattwire commented:
test this please

mattwire commented:
test this please

eileenmcnaughton commented:
This has good test cover & most review has been responded to - there is just one comment by @michaelmcandrew that I think might need more response.


If you were opening this today I think we would insist it was apiv4 - @colemanw that raises the same 'how would it look question' as another similar job api one today

mattwire commented:
@eileenmcnaughton @colemanw The issue with scheduled jobs and API4 is that we can't run API4 scheduled jobs yet because the scheduled jobs manager only works with API3 - if we could remove that blocker then I'd start writing a lot more in API4!

mattwire committed:
Add api3 metadata to validate exclude_membership_status_ids

mattwire commented:
@eileenmcnaughton Have responded - pseudoconstant does not load in API3 UI but I've added it anyway. And now validating as INT

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
OK sounds good - I've removed the needs-documentation flag as I see that as optional now as I really want to pivot to a v4 focus. To be clear I don't want to discourage you - but it's less compulsory now as we try to get away from apiv3 I think

freeform-sg commented:
Hi there, thank you so much for this. I came upon this wondering why my membership statuses weren't updating.

After testing this out, I have a couple of comments: 

1. the documentation does not mention that parameters are now available for this job
2. the parameter "exclude_membership_status_ids" implies by the name that it accepts an array (plural ids) but only accepts a single integer

I see you wanted to focus on api4 - is there somewhere else I can monitor the progression of this? I was having trouble finding anything related to this job.

Thanks!
